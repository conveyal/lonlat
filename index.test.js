/* globals describe, expect, it */

const ll = require('./')

const lat = 38.13234
const lon = 70.01232
const lonlat = {lon, lat}
const point = {x: lon, y: lat}
const coordinates = [lon, lat]
const str = `${lon},${lat}`
const latlng = {lat, lng: lon}

describe('lonlat', () => {
  describe('normalization', () => {
    const testCases = [{
      calculated: ll(lonlat),
      description: 'Object with lon and lat keys'
    }, {
      calculated: ll(point),
      description: 'Object with x and y keys'
    }, {
      calculated: ll(coordinates),
      description: 'Array of lon and lat'
    }, {
      calculated: ll(str),
      description: 'String with comma separating lon and lat'
    }]

    testCases.forEach((test) => {
      it(`should normalize from ${test.description}`, () => {
        expect(test.calculated).toEqual(lonlat)
      })
    })
  })

  describe('translations', () => {
    const testCases = [{
      expected: coordinates,
      method: 'toCoordinates'
    }, {
      expected: point,
      method: 'toPoint'
    }, {
      expected: str,
      method: 'toString'
    }]

    testCases.forEach((test) => {
      it(`should translate using ${test.method}`, () => {
        expect(ll[test.method](lonlat)).toEqual(test.expected)
      })
    })
  })

  describe('known type parsing', () => {
    const testCases = [{
      calculated: ll.fromLatlng(latlng),
      description: 'Object with lng and lat keys'
    }, {
      calculated: ll.fromCoordinates(coordinates),
      description: 'Array of lon and lat'
    }, {
      calculated: ll.fromPoint(point),
      description: 'Object with x and y keys'
    }, {
      calculated: ll.fromString(str),
      description: 'String with comma separating lon and lat'
    }]

    testCases.forEach((test) => {
      it(`should specifically parse from ${test.description}`, () => {
        expect(test.calculated).toEqual(lonlat)
      })
    })
  })

  describe('invalid coordinates', () => {
    const badCoords = [
      '-999,999',
      '0,999'
    ]

    badCoords.forEach((data) => {
      it(`should throw error when parsing: ${JSON.stringify(data)}`, () => {
        expect(() => ll(data)).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe('issues', () => {
    it('#3 - Does not parse coordinates with 0 for lat or lon', () => {
      expect(ll({ lat: 0, lng: 0 })).toEqual({ lat: 0, lon: 0 })
    })
  })
})
