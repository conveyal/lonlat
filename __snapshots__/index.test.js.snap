// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lonlat errors invalid coordinates should throw error when parsing: "-999,999" 1`] = `"Invalid longitude value: -999"`;

exports[`lonlat errors invalid coordinates should throw error when parsing: "0,999" 1`] = `"Invalid latitude value: 999"`;

exports[`lonlat errors invalid coordinates should throw error when parsing: {"latitude":1234,"lng":1} 1`] = `"Invalid latitude value: 1234"`;

exports[`lonlat errors invalid coordinates should throw error when parsing: {} 1`] = `"Invalid longitude value: undefined"`;

exports[`lonlat errors invalid coordinates should throw error when parsing: undefined 1`] = `"Value must not be null or undefined."`;

exports[`lonlat errors toLeaflet should throw error if leaflet is not loaded 1`] = `"Leaflet not found."`;
